aiida/tools/graph/age_entities.py

```diff
@@ -0,0 +1,487 @@
+        new.keyset = self.keyset.union(other.keyset)
+        return new
+
+    def __iadd__(self, other):
+        """Addition inplace (self += other)"""
+        self._check_self_and_other(other)
+        self.keyset = self.keyset.union(other.keyset)
+        return self
+
+    def __sub__(self, other):
+        """Subtraction (return = self - other): defined as the set-difference"""
+        self._check_self_and_other(other)
+        new = self.get_template()
+        new.keyset = self.keyset.difference(other.keyset)
+        return new
+
+    def __isub__(self, other):
+        """Subtraction inplace (self -= other)"""
+        self._check_self_and_other(other)
+        self.keyset = self.keyset.difference(other.keyset)
+        return self
+
+    def __len__(self):
+        return len(self.keyset)
+
+    def __repr__(self):
+        return '{{{}}}'.format(','.join(map(str, self.keyset)))
+
+    def __eq__(self, other):
+        return self.keyset == other.keyset
+
+    def __ne__(self, other):
+        return not self == other
+
+    def set_entities(self, new_entitites):
+        """
+        Replaces contained set with the new entities.
+
+        :param new_entities: entities which will replace the ones contained
+            by the EntitySet. Must be an AiiDA instance (Node or Group) or
+            an appropriate identifier (ID).
+        """
+        self.keyset = set(map(self._check_input_for_set, new_entitites))
+
+    def add_entities(self, new_entitites):
+        """
+        Add new entitities to the existing set of self.
+
+        :param new_entities: an iterable of new entities to add.
+        """
+        self.keyset = self.keyset.union(list(map(self._check_input_for_set, new_entitites)))
+
+    def copy(self):
+        """Create new instance with the same defining attributes and the same keyset."""
+        new = self.get_template()
+        new.keyset = self.keyset.copy()
+        return new
+
+    def empty(self):
+        """Resets the contained set to be an empty set"""
+        self.keyset = set()
+
+
+class AiidaEntitySet(AbstractSetContainer):
+    """Extension of AbstractSetContainer
+
+    This class is used to store `graph nodes` (aidda nodes or aiida groups).
+    """
+
+    def __init__(self, aiida_cls):
+        """Initialization method
+
+        :param aiida_cls: a valid AiiDA ORM class (Node or Group supported).
+        """
+        super().__init__()
+        if not aiida_cls in VALID_ENTITY_CLASSES:
+            raise TypeError('aiida_cls has to be among:{}'.format(VALID_ENTITY_CLASSES))
+        self._aiida_cls = aiida_cls
+        self.keyset = set()
+        self._identifier = 'id'
+        self._identifier_type = int
+        # NOTE: Is the identifier type really necessary? In the end, the QueryBuilder
+        # would return always the same type, wouldn't it? For future: Customize both
+        # _identifier and _identifier_type (uuid or name (for groups) could also work).
```

Comment:
Is this already clear now? If so, I would simplify the code now and not later

Reply:
I'm not entirely sure if it would not be eventually convenient to have different identifiers. By "simplify the code" you mean removing this properties and hard-coding the checks to be performed directly with id?

Reply:
I think that this should never really be customizable. How entities are referred to internally by the traverser is of no interest to the user, maybe only through what type the original entities are identified. The traversal is always done within a single database and all entity types have their own set and a pk, meaning that a pk can always be used. In principle there is then no need to make this configurable. I notice that there currently is no public API anyway, so I propose we keep it for now since removing it might disrupt this PR too much, but then open an issue to remove this at a later point, which should not require deprecation because it concerns only purely internal code.